rasa/core/policies/ted_policy.py

```diff
@@ -0,0 +1,642 @@
+        self,
+        training_trackers: List[DialogueStateTracker],
+        domain: Domain,
+        **kwargs: Any,
+    ) -> None:
+        """Train the policy on given training trackers."""
+
+        # set numpy random seed
+        np.random.seed(self.config[RANDOM_SEED])
+
+        # dealing with training data
+        training_data = self.featurize_for_training(training_trackers, domain, **kwargs)
+
+        self._label_data = self._create_label_data(domain)
+
+        # check if number of negatives is less than number of label_ids
+        logger.debug(
+            f"Check if num_neg {self.config[NUM_NEG]} is smaller "
+            f"than number of label_ids {domain.num_actions}, "
+            f"else set num_neg to the number of label_ids - 1."
+        )
+        self.config[NUM_NEG] = min(self.config[NUM_NEG], domain.num_actions - 1)
+
+        # extract actual training data to feed to model
+        model_data = self._create_model_data(training_data.X, training_data.y)
+        if model_data.is_empty():
+            logger.error(
+                f"Can not train '{self.__class__.__name__}'. No data was provided. "
+                f"Skipping training of the policy."
+            )
+            return
+
+        # keep one example for persisting and loading
+        self.data_example = {k: [v[:1] for v in vs] for k, vs in model_data.items()}
+
+        self.model = TED(
+            model_data.get_signature(),
+            self.config,
+            isinstance(self.featurizer, MaxHistoryTrackerFeaturizer),
+            self._label_data,
+        )
+
+        self.model.fit(
+            model_data,
+            self.config[EPOCHS],
+            self.config[BATCH_SIZES],
+            self.config[EVAL_NUM_EXAMPLES],
+            self.config[EVAL_NUM_EPOCHS],
+            batch_strategy=self.config[BATCH_STRATEGY],
+        )
+
+    def continue_training(
+        self,
+        training_trackers: List[DialogueStateTracker],
+        domain: Domain,
+        **kwargs: Any,
+    ) -> None:
+        """Continue training an already trained policy."""
+
+        batch_size = kwargs.get("batch_size", 5)
+        epochs = kwargs.get("epochs", 50)
+
+        training_data = self._training_data_for_continue_training(
+            batch_size, training_trackers, domain
+        )
+
+        model_data = self._create_model_data(training_data.X, training_data.y)
+
+        self.model.fit(
+            model_data,
+            epochs,
+            [batch_size],
+            self.config[EVAL_NUM_EXAMPLES],
+            self.config[EVAL_NUM_EPOCHS],
+            batch_strategy=self.config[BATCH_STRATEGY],
+        )
+
+    def predict_action_probabilities(
+        self, tracker: DialogueStateTracker, domain: Domain
+    ) -> List[float]:
+        """Predict the next action the bot should take.
+
+        Return the list of probabilities for the next actions.
+        """
+        if self.model is None:
+            return [0.0] * domain.num_actions
```

Comment:
I think this is also a common check / calculation for all policies right? Could be moved out / be a function with a name